### PR TITLE
Add entitlement table and sso stripe feature

### PR DESCRIPTION
**Pull request by anamarn**
**TLDR**

Added Billing Entitlement table, based on stripe customer.ActiveEntitlements webhook event. In this table it has a key value pair with each key being the stripe feature lookup key and the value a boolean. We use this table in order to see if SSO or other feaures are enabled by workspace.

**In order to test: twenty-server**

Billing:

- Set IS_BILLING_ENABLED to true
- Add your BILLING_STRIPE_SECRET and BILLING_STRIPE_API_KEY
- Add your BILLING_STRIPE_BASE_PLAN_PRODUCT_ID (use the one in testMode > Base Plan)

Auth:

- Set AUTH_SSO_ENABLED to true
- Set your ACCESS_TOKEN_SECRET, LOGIN_TOKEN_SECRET, REFRESH_TOKEN_SECRET and FILE_TOKEN_SECRET
- Set IS_SSO_ENABLED feature flag to true

Stripe Webhook: 

- Authenticate with your account in the stripe CLI
- Run the command: stripe listen --forward-to http://localhost:3000/billing/webhooks

Migration:

- npx nx typeorm -- migration:run -d src/database/typeorm/core/core.datasource.ts

**In order to test: twenty site**

- Buy a subscription (you can use the card 4242...42 with expiration date later in the future)
- Go to SSO and create an OICD subscription
- Change the value in the entitlement table in order to put it in false
- An error should occur saying that the current workspace has no entitlement

**Considerations**

The data from the Entitlement table is updated based on the stripe webhook responses, and we use the customerActiveEntitlemet response to update the info on the table, however this event doesnt have the metadata containing the workspaceId. Because we cannot control at wich order the webhook send events, we force a server error if the entitlements are updated before the BillingSubscription. Stripe resends the event based on a exponential backoff (for more info see https://docs.stripe.com/webhooks#retries ) because we are in test mode Stripe retries three times over a few hours. So if the BillingEntitlement is not updated it is completely normal and it will be updated when stripe resends the event.

### CODE REVIEW

File: packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts [hunk 1]
```diff
@@ -63,6 +62,17 @@ export class BillingController {
         event.data,
       );
     }
+    if (event.type === WebhookEvent.CUSTOMER_ACTIVE_ENTITLEMENT) {
+      try {
+        await this.billingWehbookService.processCustomerActiveEntitlement(
+          event.data,
+        );
+      } catch (error) {
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts:70]: we should catch customException and not generic error. see sso.exception.ts

File: packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts [hunk 2]
```diff
@@ -63,6 +62,17 @@ export class BillingController {
...
+        res.status(500).end();
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts:71]: catch (BillingCustomerNotFoundException) { 404)
catch(error) { 500 }

File: packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts [hunk 3]
```diff
@@ -63,6 +62,17 @@ export class BillingController {
...
+
+        return;
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts:73]: might be useless

File: packages/twenty-server/src/engine/core-modules/billing/dto/checkout-session.input.ts [hunk 4]
```diff
@@ -3,7 +3,7 @@ import { ArgsType, Field } from '@nestjs/graphql';
 import { IsNotEmpty, IsOptional, IsString } from 'class-validator';
 import Stripe from 'stripe';
 
-import { SubscriptionInterval } from 'src/engine/core-modules/billing/entities/billing-subscription.entity';
+import { SubscriptionInterval } from 'src/engine/core-modules/billing/enums/subcription-interval.enum';
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/dto/checkout-session.input.ts:6]: prefix with billing 
billing-subscription-interval.enum

File: packages/twenty-server/src/engine/core-modules/billing/dto/product.input.ts [hunk 5]
```diff
@@ -2,7 +2,7 @@ import { ArgsType, Field } from '@nestjs/graphql';
 
 import { IsNotEmpty, IsString } from 'class-validator';
 
-import { AvailableProduct } from 'src/engine/core-modules/billing/interfaces/available-product.interface';
+import { AvailableProduct } from 'src/engine/core-modules/billing/enums/available-product.enum';
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/dto/product.input.ts:5]: same here

File: packages/twenty-server/src/engine/core-modules/billing/enums/webhook-events.enum.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+export enum WebhookEvent {
+  CUSTOMER_SUBSCRIPTION_CREATED = 'customer.subscription.created',
+  CUSTOMER_SUBSCRIPTION_UPDATED = 'customer.subscription.updated',
+  CUSTOMER_SUBSCRIPTION_DELETED = 'customer.subscription.deleted',
+  SETUP_INTENT_SUCCEEDED = 'setup_intent.succeeded',
+  CUSTOMER_ACTIVE_ENTITLEMENT = 'entitlements.active_entitlement_summary.updated',
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/enums/webhook-events.enum.ts:6]: CUSTOMER_ACTIVE_ENTITLEMENT_UDPATED

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-subscription.service.ts [hunk 7]
```diff
@@ -148,6 +97,23 @@ export class BillingSubscriptionService {
     }
   }
 
+  async getWorkspaceEntitlementByKey(
+    workspaceId: string,
+    lookupKey: FeatureStripeLookupKey,
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-subscription.service.ts:102]: key: BillingEntitlementKey

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-subscription.service.ts [hunk 8]
```diff
@@ -148,6 +97,23 @@ export class BillingSubscriptionService {
...
+  ) {
+    const entitlement = await this.billingEntitlementRepository.findOneBy({
+      workspaceId,
+      key: lookupKey,
+      value: true,
+    });
+
+    if (!entitlement?.value) {
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-subscription.service.ts:110]: split to be fully explicit
if (entitlement === undefined) { return false; }
return entitlement.value

File: packages/twenty-server/src/engine/core-modules/billing/enums/feature-stripe-lookup-key.enum.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export enum FeatureStripeLookupKey {
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/enums/feature-stripe-lookup-key.enum.ts:1]: see comment below
SSO

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook.service.ts [hunk 10]
```diff
@@ -96,4 +98,34 @@ export class BillingWebhookService {
       });
     }
   }
+
+  async processCustomerActiveEntitlement(
+    data: Stripe.EntitlementsActiveEntitlementSummaryUpdatedEvent.Data,
+  ) {
+    const billingSubscription =
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook.service.ts:105]: const billingSubscription = find
if (undefined) { throw BillingSubscriptionNotFoundException }

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook.service.ts [hunk 11]
```diff
@@ -96,4 +98,34 @@ export class BillingWebhookService {
...
+      await this.billingSubscriptionRepository.findOneOrFail({
+        where: { stripeCustomerId: data.object.customer },
+      });
+    const workspaceId = billingSubscription.workspaceId;
+    const stripeCustomerId = data.object.customer;
+
+    const currentEntitlements = data.object.entitlements.data.map(
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook.service.ts:112]: currentEntitlement ==> activeEntitlement everywhere
currentEntitlements => activeEntitlementKeys

**Review by charlesBochet — APPROVED** [review]

**Comment by FelixMalfait** [comment]
Great work!
